### PR TITLE
Feature/filter page backend

### PR DESCRIPTION
**Pull request by akshire**
# Feature Filter Page Backend

## TL;DR
This PR introduces a filtering function on the Swipe screen, enabling users to view recipes that match the selected filters. Filters include time, difficulty, and category, making the recipe search more tailored and efficient.

## What Has Been Changed?
The filtering functionality has been added to the SwipeScreen, allowing users to filter recipes by their available time, the recipe's difficulty, and its category. 

**See video demonstration:**  

https://github.com/user-attachments/assets/6d1f99c2-4a4a-4133-99a5-134ab7e7af5d

As shown in the video, the time filter currently displays a test recipe because most recipes in the database do not have a time assigned. Their time parameter is `null`.

Filters are cumulative, meaning users can apply multiple filters simultaneously to refine their search further.

#### In `FirestoreRecipeRepository`
A new function has been introduced:
```kotlin
fun filterSearch(
      filter: Filter,
      onSuccess: (List<Recipe>) -> Unit,
      onFailure: (Exception) -> Unit,
      limit: Int,
  )
```
This function enables fetching recipes based on the `Filter` parameter, improving database querying flexibility.

#### In the `Filter` Class
Methods were updated to adopt a `null` parameter policy, which enhances the ability to search the database even when certain parameters are not set in the recipes.

#### In `fetchRandomRecipes(numberOfRecipes: Int)`
The `searchByCategory` function has been replaced with the `searchFilter` function. This modification enables the application of filters when fetching random recipes.

#### In the `FilterPage` Screen
The price filter field has been removed following team discussions. It was decided that price is not a necessary attribute for the current `Recipe` schema.

## Why Was This Change Made?
These changes empower users to filter recipes more effectively on the Swipe screen. The filtering options allow users to search for recipes based on their available time and preferences, providing a more personalized experience.

## Testing on Practical Data
- **Testing Process:**  
  Both UI and Unit tests have been conducted. However, due to the lack of recipes with populated time, difficulty, and category fields in the database, it was not possible to verify the correct functionality of combining multiple filters. 

- **Limitations:**  
  Testing revealed the need for a database with more diverse data to properly evaluate the functionality of combined filters.

### What Will Be Done Next?
1. Create Firebase queries to enable filtering with multiple criteria simultaneously.
2. Regenerate the database to include randomized values for difficulty and time attributes, ensuring better test coverage and enhancing the user experience.

## Checklist
- [x] Tests added:
    - `FilterPageTest`
    - `RecipeOverviewTest`
    - `SwipePageTest`
    - `FilterTest`
    - `RecipesViewModelTest`
    - `FirestoreRecipesRepositoryTest`
- [x] Documentation updated.
- [x] All CI checks passed.
- [x] Linked issue/feature (if applicable): #252

### CODE REVIEW

**Comment by akshire** [comment]
I would like to say that the 7 accepted issue come from #259, the issue was discussed and accepted

File: app/src/main/java/com/android/sample/model/filter/Filter.kt [hunk 1]
```diff
@@ -20,12 +17,6 @@ open class Filter(
             UNINITIALIZED_BORN_VALUE,
             UNINITIALIZED_BORN_VALUE,
             UNINITIALIZED_BORN_VALUE),
-    var priceRange: FloatRange =
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/filter/Filter.kt:23]: **Price Range Removal:**  
   Nice call removing `priceRange`—keeps the `Filter` class clean and focused. Make sure it’s not referenced anywhere else!

File: app/src/main/java/com/android/sample/model/filter/Filter.kt [hunk 2]
```diff
@@ -41,19 +32,15 @@ data class FloatRange(var min: Float, var max: Float, var minBorn: Float, var ma
   }
 
   fun update(newMin: Float, newMax: Float) {
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/filter/Filter.kt:34]: **`update` Method:**  
   Love how you simplified this! Removing unnecessary checks makes it way cleaner. Just double-check edge cases are covered in tests.

File: app/src/main/java/com/android/sample/model/filter/Filter.kt [hunk 3]
```diff
@@ -41,19 +32,15 @@ data class FloatRange(var min: Float, var max: Float, var minBorn: Float, var ma
...
-    require(newMin >= minBorn && newMax <= maxBorn) { NEW_MIN_AND_NEW_MAX_SHOULD_BE_WITHIN_RANGE }
-    min = newMin.roundToInt().toFloat()
-    max = newMax.roundToInt().toFloat()
+    require(newMin <= newMax) { MIN_SHOULD_NOT_BE_GREATER_THAN_MAX }
+    min = newMin
+    max = newMax
   }
 
-  fun isLimited() = min == minBorn && max == maxBorn
+  fun isLimited() =
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/filter/Filter.kt:40]: **`isLimited` Logic:**  
   Great work simplifying the logic here! Maybe toss in a quick comment for clarity—it’s a bit tricky at first glance.

File: app/src/main/java/com/android/sample/model/filter/FilterPageViewModel.kt [hunk 4]
```diff
@@ -9,7 +9,6 @@ interface FilterPageViewModel {
   val tmpFilter: StateFlow<Filter>
   val categories: StateFlow<List<String>>
   val timeRangeState: StateFlow<FloatRange>
-  val priceRangeState: StateFlow<FloatRange>
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/filter/FilterPageViewModel.kt:12]: LGTM! Make sure it’s not referenced anywhere else!

File: app/src/main/java/com/android/sample/model/recipe/DefaultRecipeRepository.kt [hunk 5]
```diff
@@ -100,16 +101,6 @@ class DefaultRecipeRepository(
     networkRepository.searchByCategory(category, onSuccess, onFailure, limit)
   }
 
-  /**
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/recipe/DefaultRecipeRepository.kt:103]: **Removing `listCategories`:**  
   Nice call removing `listCategories` if it’s no longer needed—keeps the repository lean and avoids redundant code. Just double-check that no other parts of the app rely on it.

File: app/src/main/java/com/android/sample/model/recipe/DefaultRecipeRepository.kt [hunk 6]
```diff
@@ -156,4 +147,13 @@ class DefaultRecipeRepository(
   override fun deleteAllDownloads(onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
     localRepository.deleteAll(onSuccess, onFailure)
   }
+
+  override fun filterSearch(
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/recipe/DefaultRecipeRepository.kt:151]: **Adding `filterSearch`:**  
   Cool addition of `filterSearch`! Throwing `UnsupportedOperationException` for now makes sense if it’s not yet implemented. Just remember to revisit this once the functionality is ready, or maybe add a TODO comment to ensure it’s tracked.

File: app/src/main/java/com/android/sample/model/recipe/RecipesRepository.kt [hunk 7]
```diff
@@ -1,78 +1,9 @@
 package com.android.sample.model.recipe
 
-/** Interface for retrieving recipe data. */
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/recipe/RecipesRepository.kt:3]: 1. **Inheritance Change:**  
   Smart move making `RecipesRepository` inherit from `RecipeNetworkRepository`—this really simplifies things and removes unnecessary duplication. Nice job cleaning up the interface!

2. **Deleted Methods:**  
   Makes total sense to remove those methods since they’re already handled by `RecipeNetworkRepository`. Less code duplication = easier maintenance. Just ensure no functionality was inadvertently affected.

File: app/src/main/java/com/android/sample/model/recipe/RecipesViewModel.kt [hunk 8]
```diff
@@ -80,6 +80,8 @@ class RecipesViewModel(
   override val categories: StateFlow<List<String>>
     get() = _categories
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/recipe/RecipesViewModel.kt:82]: **General Cleanliness:**  
   Overall, the code changes are clean and align well with the new filtering logic. It’s great to see how seamlessly the new functionality integrates into the existing ViewModel. LGTM!

1. **`isFilterUsed` StateFlow Addition:**  
   Adding `isFilterUsed` is a great touch! It’s a clean and efficient way to track whether filters are applied. Nicely implemented with `checkIfFilterUsed()` updates.

2. **Removing Price Range References:**  
   Smart move cleaning up `priceRange` references to align with the updated `Filter` class. It keeps the logic focused and consistent.

3. **Filter Updates (`updateCategory`, `updateTimeRange`, etc.):**  
   Good call ensuring `isFilterUsed` is updated after each filter change. This makes the code responsive to user actions and ensures the state reflects the current filter configuration.

4. **`fetchRandomRecipes` Logic:**  
   Replacing `searchByCategory` with `filterSearch` is a solid update. The `isFilterUsed` check is a thoughtful addition to ensure the backend query is only made when needed.

5. **Resetting Filters (`resetFilters`):**  
   Nice touch resetting `isFilterUsed` along with the filters. It ensures no residual state issues when users clear filters.

File: app/src/main/java/com/android/sample/model/recipe/networkData/FirestoreRecipesRepository.kt [hunk 9]
```diff
@@ -279,8 +286,81 @@ class FirestoreRecipesRepository(private val db: FirebaseFirestore) : RecipeNetw
         .addOnFailureListener { e -> onFailure(e) }
   }
 
-  /** Will be deleted in next PR */
-  override fun listCategories(onSuccess: (List<String>) -> Unit, onFailure: (Exception) -> Unit) {
-    throw NotImplementedError("Not implemented")
+  /**
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/recipe/networkData/FirestoreRecipesRepository.kt:289]: LGTM!

1. **Dynamic Filtering:**  
   Great use of `let` and `takeIf` for building filters dynamically—clean and readable.

2. **Validation:**  
   Nice touch with `require` to validate `limit` upfront.

3. **Randomness:**  
   Shuffling and limiting results adds a nice layer of unpredictability.

4. **Improvement:**  
   Consider breaking the filter logic into helper methods for better readability.

5. **Documentation:**  
   Clear and helpful comments—makes the code easy to follow.

File: app/src/main/java/com/android/sample/model/recipe/networkData/MealDBRecipesRepository.kt [hunk 10]
```diff
@@ -211,7 +212,16 @@ class MealDBRecipesRepository(private val client: OkHttpClient) : RecipeNetworkR
             })
   }
 
-  override fun listCategories(onSuccess: (List<String>) -> Unit, onFailure: (Exception) -> Unit) {
+  override fun filterSearch(
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/recipe/networkData/MealDBRecipesRepository.kt:215]: LGTM !

1. **`filterSearch` Addition:**  
   Nice placeholder with `UnsupportedOperationException`—good for tracking unimplemented features.

2. **`listCategories` Adjustment:**  
   Clean implementation of the `listCategories` function. Using `Request.Builder` for URL handling is straightforward and efficient.

File: app/src/main/java/com/android/sample/model/recipe/networkData/RecipeNetworkRepository.kt [hunk 11]
```diff
@@ -70,6 +71,18 @@ interface RecipeNetworkRepository {
       limit: Int = 5
   )
 
-  /** Lists all the categories in the API. */
-  fun listCategories(onSuccess: (List<String>) -> Unit, onFailure: (Exception) -> Unit)
+  /**
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/recipe/networkData/RecipeNetworkRepository.kt:74]: LGTM!

File: app/src/main/java/com/android/sample/resources/C.kt [hunk 12]
```diff
@@ -91,9 +91,7 @@ object C {
 
     object FilterPage {
       const val TIME_RANGE_MIN = 0f
-      const val TIME_RANGE_MAX = 200f
-      const val PRICE_RANGE_MIN = 0f
-      const val PRICE_RANGE_MAX = 100f
+      const val TIME_RANGE_MAX = 1440f // 24h*60min
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/resources/C.kt:94]: LGTM!

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt [hunk 13]
```diff
@@ -184,6 +185,23 @@ fun FilterBox(
       }
 }
 
+/**
+ * Composable function to reformat the time.
+ *
+ * @param time The time to reformat in minutes.
+ */
+@Composable
+fun reformatTime(time: Float): String {
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:194]: The changes in this file are solid and well-implemented. There’s only one thing to refactor: replacing the hardcoded `60` in `reformatTime` with the `MINUTES_PER_HOUR` constant from `C.kt` for better maintainability.

File: app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt [hunk 14]
```diff
@@ -1,6 +1,7 @@
 package com.android.sample.ui.swipePage
 
 import android.annotation.SuppressLint
+import android.util.Log
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt:4]: LGTM!

1. **`checkIfFilterUsed` Function:**  
   Nice addition! Centralizing the logic to determine if filters are active simplifies the code and reduces redundancy.

2. **`FilterChip` Updates:**  
   Adding the `checkIfFilterUsed` call within `onDelete` actions is a good improvement. It ensures filters are reset only when needed and avoids unnecessary updates.

3. **Code Cleanup:**  
   Using `checkIfFilterUsed` for all filter-related chips (time, difficulty, category) creates consistency and improves readability.

File: app/src/test/java/com/android/sample/model/filter/FilterTest.kt [hunk 15]
```diff
@@ -8,9 +8,9 @@ import org.junit.Test
 class FilterTest {
```
Comment (2Fick) [inline app/src/test/java/com/android/sample/model/filter/FilterTest.kt:8]: LGTM

File: app/src/test/java/com/android/sample/model/recipe/RecipesViewModelTest.kt [hunk 16]
```diff
@@ -113,11 +113,11 @@ class RecipesViewModelTest {
     assertThat(recipesViewModel.loading.value, `is`(false)) // Check loading is false after fetching
   }
 
+  @OptIn(ExperimentalCoroutinesApi::class)
   @Test
-  fun fetchByCategoriesRecipesHandlesFailure() {
+  fun fetchByCategoriesRecipesHandlesFailure() = runTest {
```
Comment (2Fick) [inline app/src/test/java/com/android/sample/model/recipe/RecipesViewModelTest.kt:118]: Well done

**Review by 2Fick — CHANGES_REQUESTED** [review]
You've done an awesome job with this PR, buddy! The filtering logic is clean and well-organized, and I can see how much thought you put into making everything reusable and consistent across the app.  There's great attention to detail, and the changes feel seamless with the existing code. Just that some smalls refactor, and it's golden! Keep up the great work! 🚀

**Comment by 2Fick** [comment]
Hey, I noticed a bug when applying filters—the app crashes due to an `UnsupportedOperationException`. Based on the logcat, it seems the `filterSearch` function in `DefaultRecipeRepository` is throwing the exception. Did you had any idea of this issue ? 

This should be fixed also. Don't hesitate to ask for help if needed

**Review by 2Fick — CHANGES_REQUESTED** [review]
Hey, I noticed a bug when applying filters—the app crashes due to an `UnsupportedOperationException`. Based on the logcat, it seems the `filterSearch` function in `DefaultRecipeRepository` is throwing the exception. Did you had any idea of this issue ? 

This should be fixed also. Don't hesitate to ask for help if needed

**Comment by akshire** [comment]
Yeah thanks for reporting that! It is fixed now! 😄

File: app/src/main/java/com/android/sample/model/recipe/DefaultRecipeRepository.kt (hunk 6 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/model/recipe/DefaultRecipeRepository.kt:151]: Ok this was the source of the crash therefore it is implemented now! 😄

File: app/src/main/java/com/android/sample/model/recipe/RecipesRepository.kt (hunk 7 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/model/recipe/RecipesRepository.kt:3]: Hey I think it should be ok. I ran all the tests and everything is good! 😃

File: app/src/main/java/com/android/sample/model/recipe/RecipesViewModel.kt (hunk 8 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/model/recipe/RecipesViewModel.kt:82]: Thanks m8! 👍

File: app/src/main/java/com/android/sample/model/filter/Filter.kt (hunk 2 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/model/filter/Filter.kt:34]: All the essential tests have been retained. The only changes involved removing edge cases, which are now supported directly by this function, enhancing its usability and coverage.

File: app/src/main/java/com/android/sample/model/recipe/networkData/FirestoreRecipesRepository.kt (hunk 9 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/model/recipe/networkData/FirestoreRecipesRepository.kt:289]: Thanks! 😄

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt (hunk 13 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:195]: Nice catch! I ll do it right away! 😮

File: app/src/main/java/com/android/sample/model/filter/Filter.kt (hunk 3 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/model/filter/Filter.kt:41]: It is done!

File: app/src/main/java/com/android/sample/model/filter/Filter.kt (hunk 1 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/model/filter/Filter.kt:23]: I made an attempt! As of now, apart from its mention in the Recipe, it doesn’t appear anywhere else

**Review by 2Fick — APPROVED** [review]
It seems now your feature is working perfectly. Good job for the work done and thank you for your effort, dude! This PR is now ready to merge !
